plugins/lease_time/plugin.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2018-present the CoreDHCP Authors. All rights reserved
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package leasetime
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/coredhcp/coredhcp/handler"
+	"github.com/coredhcp/coredhcp/logger"
+	"github.com/coredhcp/coredhcp/plugins"
+	"github.com/insomniacslk/dhcp/dhcpv4"
+	"github.com/insomniacslk/dhcp/dhcpv6"
+)
+
+var log = logger.GetLogger("plugins/lease_time")
+
+func init() {
+	plugins.RegisterPlugin("lease_time", setupLeaseTime6, setupLeaseTime4)
+}
+
+var (
+	v6LeaseTime time.Duration
+	v4LeaseTime time.Duration
+)
+
+// Handler6 handles DHCPv6 packets for the lease_time plugin.
+func Handler6(req, resp dhcpv6.DHCPv6) (dhcpv6.DHCPv6, bool) {
+	log.Warning("not currently supported for IPv6")
+	return resp, false
+}
+
+// Handler4 handles DHCPv4 packets for the lease_time plugin.
+func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
+	if req.OpCode != dhcpv4.OpcodeBootRequest {
+		log.Warningf("not a BootRequest, ignoring")
+		return resp, false
+	}
+	resp.Options.Update(dhcpv4.OptIPAddressLeaseTime(v4LeaseTime))
```

Comment:
So this *overwrites* the option all the time ? What about previous plugins that have already made decisions on lease times ? 
If the option is already present, I think this plugin should not touch it. It's useful when the assignment plugin doesn't have a strong opinion on lease times (because it's stateless / reservation based, like `file` or `redis`), but if the assignment plugin handles leases and keeps track of their state we can't go overwriting the values outside of them

Reply:
Yes it would overwrite all the time.  My thought was that it's behavior would be dependent on the order in which it was specified in the plugin list, ie if you put it first it's a default or if you put it after a plugin that sets lease time then it's an overwrite.  I can see utility in both.  Do you think I should make it something like "- lease_time: duration overwrite" with overwrite being a bool?

Reply:
Relying on the ordering seems fine, although a bit surprising/dangerous, and needs to be well documented then.
However, I really can't see a situation where overwriting is a good idea. Do you have one ? We can always add the option to overwrite in the future when we're sure there's a usecase for it

Reply:
The main reason I can think of you would ever want to overwrite is that you want to change every lease time to a short interval for maintenance purposes and not have to touch/understand every plugin.   Personally I think we should either have it overwrite and document it or I should add a flag for overwrite - that way users could get the most functionality out of it without someone possibly needing to make an overwrite_lease_time plugin down the road.  I'm fine with whatever you guys think though.

Reply:
For example when I place CoreDHCP into pre-production on some real life traffic I will first set the existing DHCP server to have a short like 10 minute lease time.  That way if there are issues it will only be a few minutes before the clients make a request that could (hopefully) fix the issue.  If I have my static leases all in redis with each one having it's own lease time then I would have to update each one to the shorter time and then change it back later after testing.  Again probably not a huge use case out there but nice to have the flexibility to handle it for those that might.

Reply:
In this direction, it works. The problem is if you override the lease time advertised to the client with a longer lease, then redis will expire the lease before the client, and everything breaks.

Maybe only allow overriding when it lowers the lease time then ? That's at least safe

Reply:
I must be missing something...if the lease_time plugin is say first in the list of plugins and it sets the lease time to 1m and then redis gets the field to set it to 2m then the lease time would be 2 minutes.   If redis plugin is first and sets it to 1m and then the lease_time sets it to 2m then the lease time will be 2m.   Redis is not keeping any concept of state like lease expiration time because it's purely static.   I do see your point with another plugin that might keep track of expiration.   I will make another commit shortly with an idea for a solution.

Reply:
With my latest commit configuring the plugin with "- lease_time 300s" will only set the lease time if it hasn't already been set.  If you configure as "- lease_time 300s overwrite" then it will set it regardless.

Reply:
> Redis is not keeping any concept of state like lease expiration time because it's purely static

In that kind of case, the redis plugin should not set an expiration time at all, because since it's all static, it doesn't care; then this plugin doesn't need to override anything. So for the pre-production usecase you mentioned before there's no need to overwrite. 
If however the lease expiration is stored in redis, then extending it causes problems since the value stored in redis is different from the one advertised to the client

Reply:
The lease expiration is not stored in redis so there will be no issue with that.  The redis plugin will set the lease time to the field named leasetime if it exists in the hash for a mac.  So in my use case, for example, lets say there are 5000 mac address hashes with a lease time of 30 minutes and 5000 with a lease time of 4 hours.  Now I want to do some testing/changing of the DHCP server, etc. so I want to temporarily set the lease time for everything to 5 minutes (and wait until the 4 hours are up so everything is requesting at 5 minute intervals).  The overwrite option would allow me to do that without having to update 10000 records in redis and change them back when I'm done.

Reply:
OK, I get it now. I'm not sure I understand why the redis plugin is architected that way, but that's besides the point.

We'll need a way to support both this usecase and the stateful leases usecase in the future, but for now (at least for me) this looks reasonable.

Reply:
> I'm not sure I understand why the redis plugin is architected that way, but that's besides the >point.
Do you mean why the redis plugin supports allowing you to optionally set a lease time per mac address or why it doesn't keep an expiration time?

For the stateful use case the default for this lease_time plugin should work fine....it won't set it if the option already exists.